### PR TITLE
deployment: Handle Progressing UnknownDisruption from NewReplicaSetAvailable

### PR DESCRIPTION
**Pull request by wking**
Seen on a 0.19 cluster:

```console
$ kubectl cordon control-plane-0
$ kubectl cordon control-plane-1
$ kubectl cordon control-plane-2
$ kubectl	-n openshift-cloud-credential-operator delete pods --all
...wait...
$ kubectl -n openshift-cloud-credential-operator get pods
NAME                                         READY   STATUS    RESTARTS   AGE
cloud-credential-operator-77cf8bfb6c-t6h8n   0/2     Pending   0          25m
$ kubectl -n openshift-cloud-credential-operator get replicasets
NAME                                   DESIRED   CURRENT   READY   AGE
cloud-credential-operator-77cf8bfb6c   1         1         0       89m
$ kubectl -n openshift-cloud-credential-operator get -o yaml replicaset cloud-credential-operator-77cf8bfb6c | grep -A100 '^status:'
status:
  fullyLabeledReplicas: 1
  observedGeneration: 1
  replicas: 1
$ kubectl -n openshift-cloud-credential-operator get replicasets
NAME                                   DESIRED   CURRENT   READY   AGE
cloud-credential-operator-77cf8bfb6c   1         1         0       90m
$ kubectl -n openshift-cloud-credential-operator get -o yaml deployment cloud-credential-operator | grep -A100 '^status:'
status:
  conditions:
  - lastTransitionTime: "2020-08-06T23:43:52Z"
    lastUpdateTime: "2020-08-06T23:44:23Z"
    message: ReplicaSet "cloud-credential-operator-77cf8bfb6c" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2020-08-07T00:33:26Z"
    lastUpdateTime: "2020-08-07T00:33:26Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  observedGeneration: 1
  replicas: 1
  unavailableReplicas: 1
  updatedReplicas: 1
```

With this commit, lacking available, updated pods for long enough will lead to `ProgressDeadlineExceeded`, instead of having the deployment controller falsely claiming its happy-state `Progressing=True` `NewReplicaSetAvailable` forever.

Also transition from `NewReplicaSetAvailable` to `ReplicaSetUpdated` if we lack an new/updated ReplicaSet completely.  If something removes our updated ReplicaSet and we cannot create it, or we never created it to begin with, we want to eventually end up in `ProgressDeadlineExceeded`, possibly after passing through `ReplicaSetCreateError`.

I think the happy-state `NewReplicaSetAvailable` condition should be `Progressing=False`, because we've arrived at our goal, and are no longer walking towards it.  But I'm leaving that alone for now to limit the scope of this pivot.

Not sure if this deserves a release note or not.  It is user-facing, but it is a fiddly-enough corner case that I dunno if the impact is worth the trouble and possible confusion of attempting to explain it.  Took a stab below, but I'm happy to pivot to `NONE` if folks want.

```release-note
Deployments with `progressDeadlineSeconds` set will now eventually enter `ProgressDeadlineExceeded` if the new replica set is missing or insufficient, even if the deployment was previously in the happy `NewReplicaSetAvailable` state.
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
This is pretty serious (since the feature doesn't work in all cases) so release note is appropriate.

**Comment by tnozicka** [comment]
cc @kubernetes/sig-apps-pr-reviews

File: pkg/controller/deployment/progress.go [hunk 1]
```diff
@@ -93,6 +69,25 @@ func (dc *DeploymentController) syncRolloutStatus(allRSs []*apps.ReplicaSet, new
...
+			msg := fmt.Sprintf("Deployment %q waiting on an updated replica set", d.Name)
+			condition := util.NewDeploymentCondition(apps.DeploymentProgressing, v1.ConditionTrue, util.ReplicaSetUpdatedReason, msg)
+			util.SetDeploymentCondition(&newStatus, *condition)
+
+		case currentCond != nil && currentCond.Status == v1.ConditionTrue && currentCond.Reason != util.NewRSAvailableReason:
+			// Deployment is already progressing; no need to set UnknownDisruption.
+
+		default:
+			// Deployment is not complete, but is also not currently
+			// progressing or timed out.  We may have regressed, and are
```
[the remark this answers is not in the document]
Comment (wking) [reply pkg/controller/deployment/progress.go:87]: `case util.DeploymentProgressing(d, &newStatus):` earlier should have caught the cases where we were actually progressing and not yet timed out.

File: pkg/controller/deployment/progress.go [hunk 2]
```diff
@@ -93,6 +69,25 @@ func (dc *DeploymentController) syncRolloutStatus(allRSs []*apps.ReplicaSet, new
...
+			// hopefully beginning to progress back towards our reconciled
+			// state.
+			util.SetDeploymentProgressingCondition(&newStatus, d, newRS, util.UnknownDisruption)
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/controller/deployment/progress.go:90]: DeploymentCondition.Status = corev1.ConditionUnknown

**Pull request closed without merging**